### ISSUE TITLE
Scrolling with arrow keys does not always scroll to top

### ISSUE THREAD

**Issue by jswolf19** [TonyGermaneri/canvas-datagrid#228]
_I'll look into this more when I have the time, but I wanted to record it here for now._

When using the arrow keys to scroll up, it is sometimes the case that scrolling will stop at the second row instead of the first. This does not seem to be a problem with mouse-wheel scrolling, or scrolling with the scrollbar.

### Steps to reproduce the problem.
1. Create a grid with enough data to have a vertical scrollbar. (check in the horizontal direction, too)
2. Select/activate the first row.
3. Scroll down with the mouse wheel or scroll bar
4. Press the down arrow key. This should result in the second row being displayed at the top of the data viewport.
5. Press the up arrow key.

At this point the first row should be displayed at the top of the data viewport, but it is not. It does appear to become the active row.

### Specifications like the version of the project, operating system, or hardware.
Checked in Chrome and Firefox using v0.22.12

**Comment by TonyGermaneri** [comment]
Thanks @jswolf19, sounds like an off-by-one error.

The resize event sets up all the vars the key events use to determine the location of the target row and viewport.  So the error is likely in one of these two functions:

Key events: https://github.com/TonyGermaneri/canvas-datagrid/blob/master/lib/events.js#L942

Resize event: https://github.com/TonyGermaneri/canvas-datagrid/blob/master/lib/events.js#L66

I'll stick it in the stack of bugs to get fixed.  Thanks for the report.  Of course, PRs to fix the issue are more than welcome!

Cheers!
Tony

**Comment by jswolf19** [comment]
@TonyGermaneri I was able to reproduce the issue in a test. It seems that the cells of the row above the top row are considered visible and are in the visible cells list. It seems like they are there, even though they are effectively behind the header cells (you can even see the outline of the bottom of the active cell even though the cell itself is not visible). Because of this, `scrollIntoView` does not move the scroll window.

The approach that [I am experimenting with](https://github.com/jswolf19/canvas-datagrid/commit/291955d6cbc68785c3e444c98c62445de721f3eb#diff-e8acc63b1e238f3255c900eed37254b8) is to change `scrollIntoView` to check that the top left of the cell to scroll into view is not hidden behind another cell. This seems like a brute-force approach to me, though...

I feel the test for if a cell is visible or not is already rather complicated. It might be better to look at the resize event, as you mention, or it might mean that the creation of the visible cells list needs to be looked at? What are your thoughts?

**Comment by TonyGermaneri** [comment]
Agreed, the resize method is where the problem is probably solved.  Resize function does the calculations to determine the total scrollable area, that I think is the proper place to focus.  Creation of the visible cell list is based on values calculated in the resize function.

Fairly sure it has to do with this value:

https://github.com/TonyGermaneri/canvas-datagrid/blob/master/lib/events.js#L133

`self.scrollBox.height = self.height - columnHeaderCellHeight;`

`columnHeaderCellHeight` is probably getting set incorrectly.

I'm almost certain that's the place to look.

Cheers,
Tony
